File: src/ConnectionChecker.ts
export class ConnectionChecker {
  $alive = true;
  $timer?: Timer;
  constructor(
    public $interval: number,
    public SendMessage: (message: string) => void,
    public Exit: () => void,
  ) {
    this.$ResetTimer();
  }
  CleanUp() {
    clearTimeout(this.$timer);
    this.$alive = false;
    this.$timer = undefined;
    this.SendMessage = (_ignore: string) => {};
    this.Exit = () => {};
  }
  ProcessPing() {
    this.$alive = true;
    this.SendMessage('pong');
    this.$ResetTimer();
  }
  ProcessPong() {
    this.$alive = true;
  }
  $CheckConnection() {
    if (this.$alive === true) {
      this.$alive = false;
      this.SendMessage('ping');
      this.$ResetTimer();
    } else {
      this.Exit();
      this.CleanUp();
    }
  }
  $ResetTimer() {
    clearTimeout(this.$timer);
    this.$timer = setTimeout(() => {
      this.$CheckConnection();
    }, this.$interval);
  }
}

File: ExSync-Plugin/ConnectionChecker.cs
using System;
using System.Timers;

namespace ExSyncPlugin
{
  internal class ConnectionChecker
  {
    public bool _alive = true;
    public Timer _timer;
    public ConnectionChecker(int interval, Action<string> SendMessage, Action Exit)
    {
      Interval = interval;
      this.SendMessage = SendMessage;
      this.Exit = Exit;
      _ResetTimer();
    }
    public int Interval { get; set; }
    public Action<string> SendMessage { get; set; }
    public Action Exit { get; set; }
    public void CleanUp()
    {
      if (_timer != null)
      {
        _timer.Stop();
        _timer.Dispose();
      }
      _alive = false;
      _timer = null;
      SendMessage = (string _ignore) => { };
      Exit = () => { };
    }
    public void ProcessPing()
    {
      _alive = true;
      SendMessage("pong");
      _ResetTimer();
    }
    public void ProcessPong()
    {
      _alive = true;
    }
    public void _CheckConnection()
    {
      if (_alive == true)
      {
        _alive = false;
        SendMessage("ping");
        _ResetTimer();
      }
      else
      {
        Exit();
        CleanUp();
      }
    }
    public void _ResetTimer()
    {
      if (_timer != null)
      {
        _timer.Stop();
        _timer.Dispose();
      }
      _timer = new Timer(Interval);
      _timer.Elapsed += (sender, e) =>
      {
        _CheckConnection();
      };
      _timer.AutoReset = false;
      _timer.Start();
    }
  }
}
